federation/plugin/pkg/admission/schedulingpolicy/query.go

```diff
@@ -0,0 +1,190 @@
+type policyEngineQuery struct {
+	client       *rest.RESTClient
+	retryBackoff time.Duration
+	user         user.Info
+	obj          runtime.Object
+	gvk          schema.GroupVersionKind
+}
+
+// newPolicyEngineQuery returns a policyEngineQuery that can be executed.
+func newPolicyEngineQuery(client *rest.RESTClient, retryBackoff time.Duration, user user.Info, obj runtime.Object, gvk schema.GroupVersionKind) *policyEngineQuery {
+	return &policyEngineQuery{
+		client:       client,
+		retryBackoff: retryBackoff,
+		user:         user,
+		obj:          obj,
+		gvk:          gvk,
+	}
+}
+
+// Do returns the result of the policy engine query. If the policy decision is
+// undefined or an unknown error occurs, err is non-nil. Othewrise, result is
+// non-nil and contains the result of policy evaluation.
+func (query *policyEngineQuery) Do() (decision *policyDecision, err error) {
+
+	raw, err := convertObject(query.obj, query.gvk.GroupVersion())
+	if err != nil {
+		return nil, err
+	}
+
+	body := policyEngineInput{}
+	body.Input.Resource = raw
+	body.Input.User.Name = query.user.GetName()
+	body.Input.User.Groups = query.user.GetGroups()
+	body.Input.User.UID = query.user.GetUID()
+	body.Input.User.Extra = query.user.GetExtra()
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(&body); err != nil {
+		return nil, err
+	}
+
+	var result rest.Result
+
+	err = webhook.WithExponentialBackoff(query.retryBackoff, func() error {
+		result = query.client.Post().
+			SetHeader("Content-Type", "application/json").
+			Body(&buf).
+			Do()
+		return result.Error()
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return decodeResult(result)
+}
+
+// TODO(tsandall): review the request type as well as SerDes
+
+// policyEngineInput represents the input data provided to the policy engine when
+// executing a policy query.
+type policyEngineInput struct {
+	Input struct {
+		Resource interface{} `json:"resource"`
+		User     struct {
+			Name   string              `json:"name"`
+			UID    string              `json:"uid"`
+			Groups []string            `json:"groups"`
+			Extra  map[string][]string `json:"extra"`
+		} `json:"user"`
+	} `json:"input"`
+}
+
+// policyEngineResponse represents the policy engine's response to a policy
+// query.
+type policyEngineResponse struct {
+	Result *policyDecision `json:"result,omitempty"`
+}
+
+// policyDecision represents a response from the policy engine.
+type policyDecision struct {
+	Errors   []string `json:"errors,omitempty"`
+	Resource struct {
+		Metadata struct {
+			Annotations map[string]string `json:"annotations,omitempty"`
+		} `json:"metadata,omitempty"`
+	} `json:"resource,omitempty"`
+}
+
+// Error returns an error if the policy raised an error.
+func (d *policyDecision) Error() error {
+	if len(d.Errors) == 0 {
+		return nil
+	}
+	return fmt.Errorf("reason(s): %v", strings.Join(d.Errors, "; "))
+}
+
+func convertObject(obj runtime.Object, gv schema.GroupVersion) (interface{}, error) {
```

Comment:
This was needed so that the admission controller could construct the policy query as follows:

```
{
  "input": <resource>
}
```

What I think would be better is if the `<resource>` value was not encapsulated inside `{"input": ...}`.

it should be fine to remove this entirely.